### PR TITLE
feat: separate env vars for server and client

### LINKED ISSUE #675: Allow configuration of frontend web-app env vars via the main `.env` file
**Is your feature request related to a problem? Please describe.**
Right now, we only copy the `.env` file into the server directory. This means to set frontend env vars, you have to set them when invoking the Wasp CLI command. So for example, if you want to run a Wasp app on a different port, you can set `PORT=4040` in your `.env` file to get the server running on 4040, but to have the web-app send the API calls to the backend you have to invoke `start` like so: `REACT_APP_API_URL=http://localhost:4040 wasp start`

**Describe the solution you'd like**
It would be better if we could set `REACT_APP_API_URL` in the `.env` file. Or perhaps have some cleaner way to specify it in a single env var even.

### PR DESCRIPTION
**Pull request by preshetin**
# Description

Adds support for `.env` vars for frontend app. To enable this, Wasp user can add `.env.client` file which will be passed to web app as `.env` file.

**⚠️ Breaking change**: Existing Wasp users have to manually update env filename at wasp root dir from `.env` to `.env.server`.

Fixes #675 

Draft status of this PR will be removed once the following is completed:
- [ ] Check all examples for `env-sample` file and update accordingly. By now found at https://github.com/wasp-lang/wasp/blob/main/waspc/examples/todoApp/sample.env
- [ ] Update docs at https://github.com/wasp-lang/wasp/blob/main/web/docs/language/features.md#env

## Type of change

Please select the option(s) that is more relevant.

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

### CODE REVIEW

File: waspc/src/Wasp/AppSpec.hs [hunk 1]
```diff
@@ -44,7 +44,8 @@ data AppSpec = AppSpec
     migrationsDir :: Maybe (Path' Abs (Dir DbMigrationsDir)),
     -- | Absolute path to the .env file in wasp project source. It contains env variables to be
     -- provided to the server only during the development.
-    dotEnvFile :: Maybe (Path' Abs File'),
+    dotEnvServerFile :: Maybe (Path' Abs File'),
+    dotEnvClientFile :: Maybe (Path' Abs File'),
```
Comment (preshetin) [inline waspc/src/Wasp/AppSpec.hs:48]: Proper comments have to be added for each field.

**Comment by preshetin** [comment]
Hi @shayneczyzewski and @Martinsos - can some one review this? 

It still misses some doc and example updates. I added reminders on those to the TODO in the description.

**Comment by shayneczyzewski** [comment]
> Hi @shayneczyzewski and @Martinsos - can some one review this?
> 
> It still misses some doc and example updates. I added reminders on those to the TODO in the description.

Hey there @preshetin! Great work getting a PR open! Martin is on vacation this/next week, but I can give it a first pass on Monday.

Thanks, and be in touch soon.

File: waspc/src/Wasp/AppSpec.hs (hunk 1 above)
Comment (shayneczyzewski) [reply waspc/src/Wasp/AppSpec.hs:48]: Sounds good! Additionally, I wonder if it would flow slightly better having them named `serverDotEnvFile` and `clientDotEnvFile`? 🤔 Maybe we can let @Martinsos chime in there, he is the resident naming guru. :)

File: waspc/src/Wasp/Lib.hs [hunk 2]
```diff
@@ -88,11 +90,17 @@ findWaspFile waspDir = do
...
-  return $ if dotEnvExists then Just dotEnvAbsPath else Nothing
+findDotEnvServer :: Path' Abs (Dir WaspProjectDir) -> IO (Maybe (Path' Abs File'))
+findDotEnvServer waspDir = do
+  let dotEnvServerAbsPath = waspDir SP.</> [relfile|.env.server|]
+  dotEnvServerExists <- doesFileExist (SP.toFilePath dotEnvServerAbsPath)
+  return $ if dotEnvServerExists then Just dotEnvServerAbsPath else Nothing
+
+findDotEnvClient :: Path' Abs (Dir WaspProjectDir) -> IO (Maybe (Path' Abs File'))
+findDotEnvClient waspDir = do
+  let dotEnvClientAbsPath = waspDir SP.</> [relfile|.env.client|]
```
Comment (shayneczyzewski) [inline waspc/src/Wasp/Lib.hs:101]: I don't think we exported/used them like this before, but I wonder if now is a good time to have a single definition of `.env.server` and `.env.client`? We could put them in `ServerGenerator/Common.hs` and `WebAppGenerator/Common.hs` and then use those both here in `Lib.hs` as well as `ServerGenerator.hs` and `WebAppGenerator.hs`?

Comment (shayneczyzewski) [reply waspc/src/Wasp/Lib.hs:101]: Also, since `findDotEnvServer` and `findDotEnvClient` are identical (modulo the exact env file name), should we take that as a param and have a single function definition?

**Review by shayneczyzewski — COMMENTED** [review]
Great job so far, @preshetin! 🥳 This is awesome progress! I left a few questions/comments alongside the code. Additionally, I had a few other thoughts:

- It may make sense to diff the `.gitignore` files in `.wasp/out/server` and `.wasp/out/web-app`. The server version seems to include `.env` which sounds right, but web-app does not. That file is here if you want to update it: `data/Generator/templates/react-app/gitignore`
- Were you able to find anything in CRA that would allow us to change that `REACT_APP_` envar prefix?

Thanks again for all the work here so far! Can't wait to see the comments, example apps, and docs come along as well. 🚀 👏🏻 
Shayne

File: waspc/src/Wasp/AppSpec.hs (hunk 1 above)
Comment (Martinsos) [reply waspc/src/Wasp/AppSpec.hs:48]: Hah thanks for the title @shayneczyzewski :D! I like both, as you said `serverDotEnvFile` flows a bit nicer, but on the other hand `dotEnvSomethingFile` is closer to name file has on the disk + both dot env files then share same prefix. So if I had to choose I would be slightly in favor of `dotEnvServerFile` even, but very close, so both is fine with me really.

**Review by Martinsos — COMMENTED** [review]
This is looking all good to me! As @shayneczyzewski commented, there is some douplication we might want to remove, but other than that it is looking good.
How was it doing this in Haskell, how did you find it?
TLDR; all good from my side, keep it going @preshetin and @shayneczyzewski !

File: waspc/src/Wasp/Lib.hs (hunk 2 above)
Comment (preshetin) [reply waspc/src/Wasp/Lib.hs:101]: > Also, since `findDotEnvServer` and `findDotEnvClient` are identical (modulo the exact env file name), should we take that as a param and have a single function definition?

I tried to create a general function that takes env type as a param... My Haskell skills are very limited. To do so, we have to put let's say `envType` variable inside  `[relfile|...]` expression. When I try `[relfile|envType]` it does not read variable and use its name instead.

It's pretty silly but that's what  my Haskell knowledge is

**Comment by preshetin** [comment]
Thanks @shayneczyzewski and @Martinsos for reviewing this 🙇‍♂️

My Haskell skills are miserable so even tiny refactoring may take huge amount of time.

I wanted to learn Haskell when I saw someone mention Wasp project in Prisma showcase Slack channel. Then I completed about half of the [Learn Haskell by building a blog generator](https://lhbg-book.link/01-about.html) tutorial. Thanks btw for posting this link at Wasp Discord.

**Comment by preshetin** [comment]
> * Were you able to find anything in CRA that would allow us to change that `REACT_APP_` envar prefix?

@shayneczyzewski CRA does not support changing / removing this prefix, see https://github.com/facebook/create-react-app/issues/2865. There are certainly ways to tweak things during web app generation

**Comment by shayneczyzewski** [comment]
Hi @preshetin - thanks for the replies! No worries on the Haskell refactoring timing as you learn it, it can definitely take some time to get used to. I can try to see if I can spend some time pointing you in the right direction for that function DRYing refactor. Part of it is slightly complicated by the StrongPath types. 👍🏻Keep at it though, you are doing great! :D 

Sounds good on CRA envar prefix, thanks for confirming and the link!

File: waspc/src/Wasp/Lib.hs (hunk 2 above)
Comment (shayneczyzewski) [reply waspc/src/Wasp/Lib.hs:101]: Good job so far! One way you can DRY this up would be something like:

```haskell
findDotEnvServer :: Path' Abs (Dir WaspProjectDir) -> IO (Maybe (Path' Abs File'))
findDotEnvServer waspDir = toMaybeAbsFile $ waspDir SP.</> [relfile|.env.server|]

findDotEnvClient :: Path' Abs (Dir WaspProjectDir) -> IO (Maybe (Path' Abs File'))
findDotEnvClient waspDir = toMaybeAbsFile $ waspDir SP.</> [relfile|.env.client|]

toMaybeAbsFile :: Path' Abs File' -> IO (Maybe (Path' Abs File'))
toMaybeAbsFile dotEnvFile = do
  dotEnvFileExists <- doesFileExist (SP.toFilePath dotEnvFile)
  return $ if dotEnvFileExists then Just dotEnvFile else Nothing
```

The trick is playing with the types and seeing that this has the following type:
```haskell
waspDir SP.</> [relfile|.env.foo|] :: Path' Abs File'
```

This is ok, but we kinda lose some info here w.r.t. where files live. A slight improvement might be something like:

```haskell
findDotEnvServer :: Path' Abs (Dir WaspProjectDir) -> IO (Maybe (Path' Abs File'))
findDotEnvServer waspDir = toMaybeAbsFile waspDir [relfile|.env.server|]

findDotEnvClient :: Path' Abs (Dir WaspProjectDir) -> IO (Maybe (Path' Abs File'))
findDotEnvClient waspDir = toMaybeAbsFile waspDir [relfile|.env.client|]

toMaybeAbsFile :: Path' Abs (Dir WaspProjectDir) -> Path' (Rel WaspProjectDir) File' -> IO (Maybe (Path' Abs File'))
toMaybeAbsFile waspDir dotEnvFile = do
  let dotEnvAbsFp = waspDir SP.</> dotEnvFile
  dotEnvFileExists <- doesFileExist $ SP.toFilePath dotEnvAbsFp
  return $ if dotEnvFileExists then Just dotEnvAbsFp else Nothing
```

Here, we are saying our file is _relative_ to the wasp project dir. This also helps since we could then actually move those `relfile` bits into a shared declaration, again saying they are relative to that directory, and put it in some Common.hs file:

```haskell
dotEnvClient :: Path' (Rel WaspProjectDir) File'
dotEnvClient = [relfile|.env.client|]
```

Does that kinda make sense? StrongPath and Haskell typing does take a bit of time to grok, which is totally fine! :D

Comment (Martinsos) [reply waspc/src/Wasp/Lib.hs:101]: Ah yes StrongPath is certainly a bit advanced thing, but it helps us a lot in the code - but don't worry about having a bit harder time grasping it!
Let us know if you want to do this change on your own -> if it feels too much, we will take care of it before we merge, and you can take a look at the result.

**Comment by Martinsos** [comment]
@preshetin cool that you are learning haskell!
I would recommend http://learnyouahaskell.com/ as an initial (free) resource, and building that blog post sounds like a nice project to experiment with.
Haskell is a bit scary by reputation but is actually easier than what people say + it teaches you some really interesting concepts.
Good place to ask for Haskell help is also r/haskell, but you can also ask in our wasp-lang discord, we have #haskell channel, and no question is too trivial!
Btw I hope you got LSP working for this project -> it helps a lot of and makes refactoring in Haskell really nice. If you need help with setting that up let us know although it should be pretty straightforward if you are using visual studio code.

As for this PR -> if you want to give a bit of a try to do that refactoring following @shayneczyzewski 's suggestions, go for it. If on the other hand feel that is a bit too much at the moment, that is also ok -> let us know and we will take it from here and do some final tweaks on the PR before we merge -> you can then also take a look at the result of it, to see what we did.

File: waspc/src/Wasp/Lib.hs (hunk 2 above)
Comment (preshetin) [reply waspc/src/Wasp/Lib.hs]: Thanks @shayneczyzewski for providing those chunks of code. I injected the in the codebase

Comment (preshetin) [reply waspc/src/Wasp/Lib.hs]: So @Martinsos I've impended what Shayne proposes. If you still prefer it done differently let me know

**Comment by preshetin** [comment]
Thanks @Martinsos and @shayneczyzewski for your feedback.

I removed duplication thanks to Shayne's chunks of code. Hoping someone will review it once again

**Comment by preshetin** [comment]
For now the only thing left is updating the documentation. If the rest is fine I can proceed with that. Hoping it won't take much time

File: waspc/examples/todoApp/sample.env.client [hunk 3]
```diff
@@ -0,0 +1,2 @@
+#REACT_APP_API_URL=http://localhost:4041
+#PORT=4040
```
Comment (Martinsos) [inline waspc/examples/todoApp/sample.env.client:2]: I believe PORT is not used on client, so let's drop this line!

File: waspc/examples/todoApp/sample.env.client [hunk 4]
```diff
@@ -0,0 +1,2 @@
+#REACT_APP_API_URL=http://localhost:4041
```
Comment (Martinsos) [inline waspc/examples/todoApp/sample.env.client:1]: Default for backend is 3001, so maybe let's set this to default value, that is kind of usual?

File: waspc/examples/todoApp/sample.env.server [hunk 5]
```diff
@@ -1,2 +1,3 @@
 TEST_ENV_VAR="I am test"
 DATABASE_URL=postgresql://postgres:postgres@localhost/todoapp
+#PORT=4041
```
Comment (Martinsos) [inline waspc/examples/todoApp/sample.env.server:3]: Maybe we can also set this to default value which is 3001.

File: waspc/src/Wasp/Lib.hs [hunk 6]
```diff
@@ -88,11 +92,17 @@ findWaspFile waspDir = do
...
+findDotEnvServer waspDir = toMaybeAbsFile waspDir dotEnvServer
+
+findDotEnvClient :: Path' Abs (Dir WaspProjectDir) -> IO (Maybe (Path' Abs File'))
+findDotEnvClient waspDir = toMaybeAbsFile waspDir dotEnvClient
+
+toMaybeAbsFile :: Path' Abs (Dir WaspProjectDir) -> Path' (SP.Rel WaspProjectDir) File' -> IO (Maybe (Path' Abs File'))
+toMaybeAbsFile waspDir dotEnvFile = do
+  let dotEnvAbsFp = waspDir SP.</> dotEnvFile
+  dotEnvFileExists <- doesFileExist $ SP.toFilePath dotEnvAbsFp
+  return $ if dotEnvFileExists then Just dotEnvAbsFp else Nothing
```
Comment (Martinsos) [inline waspc/src/Wasp/Lib.hs:105]: I would propose following refactoring:

```suggestion
findFileInWaspProjectDir
  :: Path' Abs (Dir WaspProjectDir)
  -> Path' (SP.Rel WaspProjectDir) File'
  -> IO (Maybe (Path' Abs File'))
findFileInWaspProjectDir waspDir file = do
  let fileAbsFp = waspDir SP.</> file
  fileExists <- doesFileExist $ SP.toFilePath fileAbsFp
  return $ if fileExists then Just fileAbsFp else Nothing
```

Basically two changes:
1. Renamed function to better reflect what it does.
2. Renamed its internals to make it aligned with the fact that it works for any file, not just env file.

**Review by Martinsos — COMMENTED** [review]
@preshetin this is it!
All is good, I only suggested one refactoring that fixes names in that one function a bit, and that is it! Once you take care of that one we can merge. Actually, we also do need to update the docs -> are you ok with moving to that step after this fix? Then we should be able to merge.

File: waspc/examples/todoApp/sample.env.client (hunk 3 above)
Comment (preshetin) [reply waspc/examples/todoApp/sample.env.client:2]: It does actually, I checked. If PORT is set in `.env.client` file and we  run `wasp start` it will use that PORT number and open a browser page at http://localhost:4040. 

I can update it to 3000 which is default client PORT value

File: waspc/examples/todoApp/sample.env.client (hunk 4 above)
Comment (preshetin) [reply waspc/examples/todoApp/sample.env.client:1]: Good idea, that'll be less confusing

File: waspc/examples/todoApp/sample.env.server (hunk 5 above)
Comment (preshetin) [reply waspc/examples/todoApp/sample.env.server:3]: I'll adjust it.

File: waspc/src/Wasp/Lib.hs (hunk 6 above)
Comment (preshetin) [reply waspc/src/Wasp/Lib.hs:105]: I agree. I also noticed that this general purpose function can be applied to any file, not just dot-env but was too shy to change proposed code.

**Comment by preshetin** [comment]
Thanks @Martinsos for the review. Sure I can do those adjustments and also update the docs. Hoping it will be within a day or two

**Comment by preshetin** [comment]
Hey @Martinsos - I've made those adjustments. The only thing left is updating docs. I believe it is in another repo. 

Still, it's better to wait with merging until docs are updated

**Comment by Martinsos** [comment]
Cool @preshetin , that concludes it, looking good!

Actually, docs are in the same repo, so the best way is to do them also in this PR.
You can find them under /web/docs/, and you will specifically be interested in /web/docs/language/features.md, I believe that is the only file you will need to change.
To check out how to build the webpage/blog/docs, check web/README.md, and ping us if anything is unclear!

File: waspc/examples/todoApp/sample.env.client (hunk 3 above)
Comment (Martinsos) [reply waspc/examples/todoApp/sample.env.client]: Ha cool I actually forgot that :)! Awesome

**Comment by preshetin** [comment]
Docs updated. If the wording needs adjusting just let me know or feel free to jump in with your commit.

If all looks good it can be merged I think.

File: web/docs/language/features.md [hunk 7]
```diff
@@ -1069,19 +1069,23 @@ Your project will likely be using environment variables for configuration, typic
...
+`.env.server` and `.env.client` files should not be commited to the version control - we already ignore it by default in the .gitignore file we generate when you create a new Wasp project via `wasp new` cli command.
+
+Variables are defined in `.env.server` or `env.client` files the same way as you would define your `.env` file in the form of `NAME=VALUE`, for example:
 ```
 DATABASE_URL=postgresql://localhost:5432
 MY_VAR=somevalue
 ```
 
-Any env vars defined in the `.env` will be forwarded to the server-side of your Wasp project and therefore accessible in your nodejs code via `process.env`, for example:
+Any env vars defined in the `.env.server` / `.env.client` files will be forwarded to the server-side / web app of your Wasp project and therefore accessible in your nodejs code via `process.env`, for example:
```
Comment (Martinsos) [inline web/docs/language/features.md:1088]:
```suggestion
Any env vars defined in the `.env.server` / `.env.client` files will be forwarded to the server-side / web app of your Wasp project and therefore accessible in your javascript code via `process.env`, for example:
```

File: web/docs/language/features.md [hunk 8]
```diff
@@ -1069,19 +1069,23 @@ Your project will likely be using environment variables for configuration, typic
...
+However, when in development, you might also need to supply certain environment variables, and to avoid doing it "manually", Wasp supports `.env` (dotenv) files where you can define environment variables that will be used during development (they will not be used during production).
 
-`.env` file has to be defined in the root of your Wasp project.
+Since environmental variables are usually different for server-side and client apps, in Wasp root directly you can create two files, `.env.server` for server-side of your Wasp project, and `.env.client` for client side (or web app) of Wasp project.
 
-`.env` file should not be commited to the version control - we already ignore it by default in the .gitignore file we generate when you create a new Wasp project via `wasp new` cli command.
+During complile, Wasp will use `.env.server` and `.env.client` files to create `.env` files and place them in server-side and web app directories.
```
Comment (Martinsos) [inline web/docs/language/features.md:1076]: I think we can skip this paragraph, it is an implementation detail they don't necessarily need to know about.

```suggestion
```

File: web/docs/language/features.md [hunk 9]
```diff
@@ -1069,19 +1069,23 @@ Your project will likely be using environment variables for configuration, typic
...
 
-Variables are defined in `.env` in the form of `NAME=VALUE`, for example:
+`.env.server` and `.env.client` files have to be defined in the root of your Wasp project.
+
+`.env.server` and `.env.client` files should not be commited to the version control - we already ignore it by default in the .gitignore file we generate when you create a new Wasp project via `wasp new` cli command.
+
+Variables are defined in `.env.server` or `env.client` files the same way as you would define your `.env` file in the form of `NAME=VALUE`, for example:
```
Comment (Martinsos) [inline web/docs/language/features.md:1082]:
```suggestion
Variables are defined in `.env.server` or `env.client` files in the form of `NAME=VALUE`, for example:
```

**Review by Martinsos — APPROVED** [review]
Great! This is looking good, I added couple minor changes to wording, will commit those now and merge!

**Review by Martinsos — APPROVED** [review]

**Comment by Martinsos** [comment]
A thought -> what if somebody has .env defined from before, updates Wasp, and doesn't realize .env doesn't do anything any more -> it might take them some time to figure out this is the problem! Maybe we should add a piece of logic that scans for `.env` file and warns that it is redundant in such case? What do you think @preshetin @shayneczyzewski ? Or should we just ignore it?

**Comment by shayneczyzewski** [comment]
> A thought -> what if somebody has .env defined from before, updates Wasp, and doesn't realize .env doesn't do anything any more -> it might take them some time to figure out this is the problem! Maybe we should add a piece of logic that scans for `.env` file and warns that it is redundant in such case? What do you think @preshetin @shayneczyzewski ? Or should we just ignore it?

Good question and reminder, @Martinsos. Given that this will break backward compatibility and some other framework norms, I see 3 options off the top of my head:

1. [Nothing] We do nothing, and just call it a breaking change (and highlight it prominently in changelog/docs)
2. [Warn] We issue a compiler warning if we see a `.env` file in the wasp project dir that it should be a client/server file instead, like you said
3. [Fix] We treat any `.env` as a `.env.server` file and "fix" it for them

I would personally probably lean toward #2. It helps guide both existing users toward the new norm, and new users who may get caught in the future. What do you think @preshetin? Great job so far BTW!

**Comment by Martinsos** [comment]
Yes, let's do (2)!
@preshetin is this something you would be ok with adding as a final step?

**Comment by preshetin** [comment]
> 1. [Nothing] We do nothing, and just call it a breaking change (and highlight it prominently in changelog/docs)
> 2. [Warn] We issue a compiler warning if we see a `.env` file in the wasp project dir that it should be a client/server file instead, like you said
> 3. [Fix] We treat any `.env` as a `.env.server` file and "fix" it for them

I would lean either to [Nothing] or [Warn]. Manipulating user's code without explicit asking feels like a bad idea to me.

As to implementation of [Warn], I'm not sure I can do it quickly. So if anyone feels to jump in I am totally fine with that!

**Comment by preshetin** [comment]
> Yes, let's do (2)! @preshetin is this something you would be ok with adding as a final step?

@Martinsos I'm afraid it'll take too much time for me. So if we want to ship it faster better for someone to jump in

**Comment by Martinsos** [comment]
No problem @preshetin , we can take care of this one! Should be relatively easy by using the function you wrote, the one finds any file in the wasp project dir. Thanks for pushing this PR so far, we will ping you once we are done and merge it!

**Comment by shayneczyzewski** [comment]
Hey @Martinsos and @preshetin, I had a few moments after wrapping some things up to come back and take a look at this. I figured the easiest way to propose the `GeneratorWarning` addition was to just make a PR based off your branch Peter.

So here it is: https://github.com/preshetin/wasp/pull/1

Please let me know if that makes sense to both of you. Thanks!
